File: python_impl/net.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Iterable, List, NewType, Tuple
import numpy as np

DataSet = NewType('DataSet', List[Tuple[np.ndarray, int]])


class Net:
    def __init__(
        self,
        n_inputs: int,
        n_outputs: int,
        hidden_layer_sizes: List[int],
    ) -> None:
        self._layer_size = [n_inputs, *hidden_layer_sizes, n_outputs] 
        self._n_layers = len(self._layer_size)
        # Note: weight matrix has dimensions (layer_size[n+1], layer_size[n])
        #       so we can obtain n+1 layer activations simply by sigmoid(np.matmul(weight[n],activ[n]) + bias[n])
        self._weights = [np.random.randn(n2, n1) 
                         for n1, n2 in zip(self._layer_size[:-1], self._layer_size[1:])]
        self._biases = [np.random.randn(n, 1) for n in self._layer_size[1:]]
        
    @staticmethod
    def _sigmoid(x: np.ndarray) -> np.ndarray:
        return 1.0 / (1.0 + np.exp(-x))
    
    def _sigmoid_derivative(self, x: np.ndarray) -> np.ndarray:
        return self._sigmoid(x) * (1 - self._sigmoid(x))
    
    def forward_pass(self, inputs: np.ndarray) -> np.ndarray:
        activation = inputs
        for b, w in zip(self._biases, self._weights):
            activation = self._sigmoid(np.matmul(w, activation) + b)
        return activation
    
    def update_parameters(self, mini_batch: DataSet, learning_rate: float) -> None:
        X = np.hstack([ex[0] for ex in mini_batch])
        y = np.hstack([ex[1] for ex in mini_batch])

        vgrad_b, vgrad_w = self._back_prop(X, y)

        self._weights = self._get_new_parameters(
            self._weights, vgrad_w, learning_rate, len(mini_batch))
        
        self._biases = self._get_new_parameters(
            self._biases, vgrad_b, learning_rate, len(mini_batch))
        
    def evaluate(self, test_set: DataSet) -> int:
        # Calculate number of correct answers
        results = [(np.argmax(self.forward_pass(data)), target) for data, target in test_set]
        return sum(int(pred == target) for pred, target in results)
    
    def _cost_derivative(self, output_activations, targets):
        # Cost for a data sample x is 0.5 * (activations(x) - target)**2
        return output_activations - targets
    
    def _get_new_parameters(
        self,
        params: List[np.ndarray],
        grad_params: List[np.ndarray],
        learning_rate: float,
        sample_size: int
    ) -> List[np.ndarray]:
        return [p - learning_rate * grad_p / sample_size for p, grad_p in zip(params, grad_params)]

    def _back_prop(self, X, y):
        # X and y are matrices with horizontally stacked examples and labels
        grad_b = [np.zeros_like(b) for b in self._biases]
        grad_w = [np.zeros_like(w) for w in self._weights]

        activations = [X]
        weighted_inputs = []
        for b, w in zip(self._biases, self._weights):
            weighted_inputs.append(np.matmul(w, activations[-1]) + b)
            activations.append(self._sigmoid(weighted_inputs[-1]))

        error = self._cost_derivative(activations[-1], y) * \
            self._sigmoid_derivative(weighted_inputs[-1])
            
        grad_b[-1] = np.expand_dims(error.sum(axis=1), axis=1)
        grad_w[-1] = np.matmul(error, activations[-2].transpose())

        # for layers L - 1 ... 2, but indexes are L - 2 ... 1
        for l in range(self._n_layers - 2, 0, -1):
            
            sig_der = self._sigmoid_derivative(weighted_inputs[l - 1])
            # weights, grad_b and grad_w lists have n_layers - 1 elements
            error = np.matmul(self._weights[l].transpose(), error) * sig_der
            grad_b[l - 1] = np.expand_dims(error.sum(axis=1), axis=1)
            grad_w[l - 1] = np.matmul(error, activations[l - 1].transpose())

        return (grad_b, grad_w)
    

class SGD:
    def __init__(
        self,
        train_set: DataSet,
        test_set: DataSet,
        mini_batch_size: int,
        epochs: int,
        learning_rate: float
    ) -> None:
        self._train_set = train_set
        self._test_set = test_set
        self._n_train = len(train_set)
        self._n_test = len(test_set)
        self._mini_batch_size = mini_batch_size
        self._epochs = epochs
        self._learning_rate = learning_rate
        
    def train(self, net: Net):
        for epoch in range(self._epochs):
            mini_batches = self._shuffle_train_set_and_make_mini_batches()
            for mini_batch in mini_batches:
                net.update_parameters(mini_batch, self._learning_rate)
            self._log_performance(net, epoch)

    def _log_performance(self, net: Net, epoch: int) -> None:
        n_correct = net.evaluate(self._test_set)
        percentage = n_correct / self._n_test
        percentage_str = "{:.2%}".format(percentage)
        print(f"Epoch {epoch}/{self._epochs}: ACC {n_correct}/{self._n_test}, {percentage_str}")    
    
    def _shuffle_train_set_and_make_mini_batches(self) -> Iterable[DataSet]:
        np.random.shuffle(self._train_set)
        return [
            self._train_set[k: k + self._mini_batch_size] for k in range(0, self._n_train, self._mini_batch_size)
        ]

File: monkey_brain/src/net.rs

use ndarray::{Array2, Array, Axis, concatenate};
use ndarray_rand::RandomExt;
use rand::seq::SliceRandom;
use rand::thread_rng;
use ndarray_rand::rand_distr::StandardNormal;

#[derive(Debug)]
pub struct Net {
    n_layers: usize,
    weights: Vec<Array2<f64>>,
    biases: Vec<Array2<f64>>
}

impl Net {
    pub fn new(
        layer_sizes: Vec<usize>,
    ) -> Net {
        let n_layers = layer_sizes.len();
        let mut weights = Vec::new();
        let mut biases = Vec::new();
        
        for layer in 1..n_layers {
            let b = Array::random((layer_sizes[layer], 1), StandardNormal);
            let w = Array::random(
                (layer_sizes[layer], layer_sizes[layer - 1]),
                StandardNormal,
            );

            biases.push(b);
            weights.push(w);
        }

        Net {
            n_layers,
            weights,
            biases
        }
    }

    fn sigmoid(x: &Array2<f64>) -> Array2<f64> {
        x.mapv(|a| 1.0 / (1.0 + (-a).exp()))
    }

    fn sigmoid_prime(x: &Array2<f64>) -> Array2<f64> {
        let sigx = Net::sigmoid(x);
        sigx.mapv(|a| 1.0 - a) * sigx 
    }

    fn cost_prime(output_activations: &Array2<f64>, targets: &Array2<f64>) -> Array2<f64> {
        // partial derivatives vector of cost w.r.t. output activations
        return output_activations - targets
    }

    pub fn forward_pass(&self, inputs: &Array2<f64>) -> Array2<f64> {
        let mut a = inputs.clone();
        for (b, w) in self.biases.iter().zip(&self.weights) {
            let z = w.dot(&a) + b;
            a = Net::sigmoid(&z);
        }
        a
    }

    pub fn update_parameters(
        &mut self, 
        mini_batch: &[(Array2<f64>, Array2<f64>)], 
        learn_rate: f64
    ) {
        let mut mb_iter = mini_batch.iter();
        let mb_first = mb_iter.next().unwrap();
        let mut x_mat = mb_first.0.clone();
        let mut y_mat = mb_first.1.clone();

        for (x, y) in mb_iter {
            x_mat = concatenate![Axis(1), x_mat, *x];
            y_mat = concatenate![Axis(1), y_mat, *y];
        }

        let (grad_b, grad_w) = self.backprop(&x_mat, &y_mat);

        let scaled_rate = learn_rate / mini_batch.len() as f64;
        for (w, gw) in self.weights.iter_mut().zip(&grad_w) {
            *w -= &(scaled_rate * gw);
        }
        for (b, gb) in self.biases.iter_mut().zip(&grad_b) {
            *b -= &(scaled_rate * gb);
        }
    }

    pub fn evaluate(&self, test_set: &[(Array2<f64>, usize)]) -> usize {
        // Calculate number of correct answers
        let mut correct = 0;
        for (x, y) in test_set.iter() {
            let out = Net::argmax(&self.forward_pass(x));
            if &out == y { correct += 1};
        }
        correct
    }

    fn argmax(arr: &Array2<f64>) -> usize {
        let mut max = f64::MIN;
        let mut i_max = 0;
        for (i, el) in arr.iter().enumerate() {
            if el > &max {
                i_max = i;
                max = *el;
            }
        }
        i_max
    }

    fn backprop(
        &self, x: &Array2<f64>, y: &Array2<f64>
    ) -> (Vec<Array2<f64>>, Vec<Array2<f64>>) {
        // Return a tuple (grad_b, grad_w) representing the
        // gradient for the cost function. 
        let mut grad_b = Vec::new();
        let mut grad_w = Vec::new();

        for (b, w) in self.biases.iter().zip(&self.weights) {
            grad_b.push(Array2::<f64>::zeros(b.raw_dim()));
            grad_w.push(Array2::<f64>::zeros(w.raw_dim()));
        }

        let mut activations = vec![x.clone()];
        let mut weighted_inputs = Vec::new();

        for (b, w) in self.biases.iter().zip(&self.weights) {
            let previous_activation = activations.last().unwrap();
            let z = w.dot(previous_activation) + b;
            activations.push(Net::sigmoid(&z));
            weighted_inputs.push(z);
        }

        // 1st equation for error in the output layer
        let error = Net::cost_prime(
            &activations.pop().unwrap(), y
        ) * Net::sigmoid_prime(weighted_inputs.last().unwrap());

        let act = activations
            .pop()
            .unwrap();
        
        *grad_w.last_mut().unwrap() = error.dot(&act.t());
        *grad_b.last_mut().unwrap() = error.sum_axis(Axis(1)).insert_axis(Axis(1));
        
        for l in (1 ..= self.n_layers - 2).rev() {
            // segond equation for error in terms of the error
            // in the next layer
            // l is the index of the next layer
            // for 3 layers, the layer idxs are 0, 1, 2
            // the weights w_1 and w_2 have indexes w[0] and w[1]
            let error = self.weights[l].t().dot(&error)
                        * Net::sigmoid_prime(&weighted_inputs[l - 1]);

            grad_b[l - 1] = error.sum_axis(Axis(1)).insert_axis(Axis(1));
      
            let act = activations
                .pop()
                .unwrap();

            grad_w[l - 1] = error.dot(&act.t());
        }
        
        (grad_b, grad_w)
    }
}


pub struct SGD {
    train_set: Vec<(Array2<f64>, Array2<f64>)>,
    test_set: Vec<(Array2<f64>, usize)>,
    mini_batch_size: usize,
    epochs: usize,
    learn_rate: f64,
}

impl SGD {
    pub fn new(
        train_set: Vec<(Array2<f64>, Array2<f64>)>,
        test_set: Vec<(Array2<f64>, usize)>,
        mini_batch_size: usize,
        epochs: usize,
        learn_rate: f64,
    ) -> SGD {
        SGD {
            train_set,
            test_set,
            mini_batch_size,
            epochs,
            learn_rate,
        }
    }

    pub fn train(&self, net: &mut Net) {
        let mut rng = thread_rng();
        let mut train_set = self.train_set.clone();

        for epoch in 0..self.epochs {
            train_set.shuffle(&mut rng);
            for mini_batch in train_set.chunks(self.mini_batch_size) {
                net.update_parameters(mini_batch, self.learn_rate)
            }
            self.log_performance(net, epoch);
        }
    }

    fn log_performance(&self, net: &Net, epoch: usize) {
        let n_correct = net.evaluate(&self.test_set);
        let n_samples = self.test_set.len();
        let percentage = n_correct as f64 / n_samples as f64;
        println!("Epoch {} / {}: ACC {} / {} = {}", epoch, self.epochs, n_correct, n_samples, percentage);
    }
}
